src/pytest_bdd/parser.py

```diff
@@ -301,119 +231,225 @@ class Step:
+        """Map a step keyword to its corresponding type.
 
-        for row in rows:
-            assert len(header) == len(row)
-            yield dict(zip(header, row))
+        Args:
+            keyword (str): The keyword for the step (e.g., 'given', 'when', 'then').
 
-    def __bool__(self) -> bool:
-        return bool(self.examples)
+        Returns:
+            Optional[str]: The type of the step, or None if the keyword is unknown.
+        """
+        return {
+            "given": GIVEN,
+            "when": WHEN,
+            "then": THEN,
+        }.get(keyword)
+
+    @staticmethod
+    def parse_steps(steps_data: list[GherkinStep]) -> list[Step]:
+        """Parse a list of step data into Step objects.
+
+        Args:
+            steps_data (List[dict]): The list of step data.
+
+        Returns:
+            List[Step]: A list of Step objects.
+        """
+        steps = []
+        current_type = None
+        for step_data in steps_data:
+            name = strip_comments(step_data.text)
+            if step_data.docString:
+                name = f"{name}\n{step_data.docString.content}"
+            keyword = step_data.keyword.lower()
+            if keyword in STEP_TYPES:
+                current_type = keyword
+            steps.append(
+                Step(
+                    name=name,
+                    type=current_type,
+                    indent=step_data.location.column - 1,
+                    line_number=step_data.location.line,
+                    keyword=step_data.keyword.title(),
+                )
+            )
+        return steps
 
+    def parse_scenario(self, scenario_data: GherkinScenario, feature: Feature) -> ScenarioTemplate:
+        """Parse a scenario data dictionary into a ScenarioTemplate object.
 
-def get_tags(line: str | None) -> set[str]:
-    """Get tags out of the given line.
+        Args:
+            scenario_data (dict): The dictionary containing scenario data.
+            feature (Feature): The feature to which this scenario belongs.
 
-    :param str line: Feature file text line.
+        Returns:
+            ScenarioTemplate: A ScenarioTemplate object representing the parsed scenario.
+        """
+        templated = bool(scenario_data.examples)
+        scenario = ScenarioTemplate(
+            feature=feature,
+            name=strip_comments(scenario_data.name),
+            line_number=scenario_data.location.line,
+            templated=templated,
+            tags=self.get_tag_names(scenario_data.tags),
+            description=textwrap.dedent(scenario_data.description),
+        )
+        for step in self.parse_steps(scenario_data.steps):
+            scenario.add_step(step)
 
-    :return: List of tags.
-    """
-    if not line or not line.strip().startswith("@"):
-        return set()
-    return {tag.lstrip("@") for tag in line.strip().split(" @") if len(tag) > 1}
+        for example_data in scenario_data.examples:
```

Comment:
scenario_data.examples may be None

Reply:
The new parser always has an empty list and never None.  Should I coerce to None in the case on an empty list, or keep the parser output as is?

Reply:
ah then let's remove `| None` from the annotation

Reply:
Done in the Gherkin parser :)